### DISCUSSION TITLE
What is the semantic of ForwardOnlyMigration?

### DISCUSSION THREAD

**Discussion by dodyg** [fluentmigrator/fluentmigrator#1542] [Q&A] [answered]
Is it supposed to throw an exception if I do `migration.exe --down 0`? I am using `3.3.1` on Postgres.

**Comment by fubar-coder** [comment, marked as answer]
Yes, it throws an exception

**Reply by jzabroski** [reply]
To add to this,  in case this question is really an[ x-y problem](https://xyproblem.info/), if you are trying to only run down migrations, you can consider a "workaround" for the exception throwing behavior by applying a Tag to the down migration and only running migrations with that tag.  Similarly, you would have to add a Tag to all ForwardOnlyMigrations...

e.g., the folllowing should work (I did not run it, just writing it from memory).

```c#
[Migration(0, "I'm a forward only migration")]
[Tags("Forward")]
public class M0 : ForwardOnlyMigration
{
  public void Up()
  {
    Create.Table("Foo").WithColumn("Id").AsInt32().NotNullable().PrimaryKey();
  }
}

[Migration(1, "I'm a down only migration")]
[Tags("Downward")]
public class M1 : Migration
{
  public void Up()
  {
    // Do nothing
  }

  public void Down()
  {
    Delete.Table("Foo");
  }
}
```

I do not recommend doing the above approach, but if you are in a jam and need an urgent workaround, the above approach to refactoring your migrations should work.  I would only advise doing this if you have literally no other options and need to deliver your software.

The reason the "correct" behavior is to throw an exception is to avoid scenarios where someone tries to `migrate down` an `Up()` migration that has no corresponding `Down()` logic.  I know in my code base I would be shocked if someone ran down, so I think it should throw, since we basically never write Down scripts where I work right now.   On the opposite end of the spectrum, I have worked at big financial institutions where it was a requirement to have rollback scripts, however.  So, it also makes sense when you think about it from the perspective of needing to dot your i's and cross your t's that every change you make has a corresponding rollback command, in the case where you execute `migrate down ` just to generate the SQL and not execute it.  This is a common scenario for large U.S. banks
